Begin adding support for domains with non-conforming blocks

kidder opened the pull request:
## Proposed changes

By non-conforming Blocks I mean that the block logical coordinates of the two Blocks on an interface between them are not related by a discrete rotation (represented by an OrientationMap which could be the identity map, i.e. no rotation).  There are two main consequences that break existing code:
- A Block may have multiple neighboring Blocks in a given Direction
- Boundary data communicated to a non-conforming neighbor can not be copied or projected onto the mortar, but instead must be interpolated.   

Specific changes in this PR:
- Add template parameter IdType to Neighbors in order to use it for neighbors of either a Block or Element
- replace BlockNeighbor with Neighbors in order to support multiple block neighbors in a given direction
- introduce InterfaceDataPolicy which will be used to determine the communication pattern for applying boundary corrections
- store the InterfaceDataPolicy in Neighbors along with the (now optional) OrientationMap 



### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- replace `BlockNeighbor<Dim>` with `Neighbors<Dim, size_t>`
- when constructing block neighbors, it now needs a `std::unordered_set<size_t>` instead of a `size_t` for the Id
- if the neighboring Blocks/Elements are conforming and aligned, you need only construct the Neighbors with the set of Ids
- if the neighboring Blocks/Elements are conforming, but not aligned, you need to construct the Neighbors with the set of Ids, `domain::InterfaceDataPolicy::OrientCopyProject`, and the OrientationMap
- if you call `Neighbors::orientation()` it now returns a `std::optional` so you instead to call `Neighbors::orientation().value_or(OrientationMap<VolumeDim>::create_aligned())` to get the previous behavior.  (In the future, the code calling this will need to be modified in order to support domains with non-conforming Blocks.)
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Add template parameter for IdType to Neighbors

This is the first step to unifying BlockNeighbor with Neighbors in
order to support spherical shells using spherical harmonics.

nilsvu commented on src/Domain/Amr/NeighborsOfParent.hpp:
```diff
@@ -33,7 +33,7 @@ namespace amr {
 /// `parent_id`, that is created from its `children_elements_and_neighbor_info`
 template <size_t VolumeDim>
-std::pair<DirectionMap<VolumeDim, Neighbors<VolumeDim>>,
+std::pair<DirectionMap<VolumeDim, Neighbors<VolumeDim, ElementId<VolumeDim>>>,
```
Suggestion: define `using ElementNeighbors<Dim> = Neighbors<Dim, ElementId<Dim>>` in `Neighbors.hpp`, and some for `BlockNeighbors`? Then we don't have to repeat this type everywhere.

nilsdeppe replied:
If we want to be able to forward declare those we need to use a strong typedef instead of a using. I have no objections to the suggestion, just expanding on it.

kidder replied:
- made and used a strong typedef for BlockNeighbors
-  did not make one for ElementNeighbors as that would require a global replacement.  If you want me to do so, I will, but in a a separate PR

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -23,7 +23,7 @@ Block<VolumeDim>::Block(
         Frame::BlockLogical, Frame::Inertial, VolumeDim>>&& stationary_map,
     const size_t id,
-    DirectionMap<VolumeDim, BlockNeighbor<VolumeDim>> neighbors,
+    DirectionMap<VolumeDim, Neighbors<VolumeDim, size_t>> neighbors,
```
Suggestion to define `using BlockNeighbors<Dim> = Neighbors<VolumeDim, size_t>`

kidder replied:
used strong typedef instead

nilsvu commented on src/Domain/Structure/Neighbors.hpp:
```diff
@@ -21,20 +19,25 @@ class ElementId;
 template <size_t VolumeDim, typename IdType = ElementId<VolumeDim>>
 class Neighbors {
+  static_assert(std::is_same_v<IdType, size_t> or
+                std::is_same_v<IdType, IdType>);
```
one `IdType` -> `ElementId`?

kidder replied:
oops

nilsvu commented on src/Domain/Amr/NewNeighborIds.cpp:
```diff
@@ -49,7 +49,8 @@ std::unordered_map<ElementId<VolumeDim>, Mesh<VolumeDim>> new_neighbor_ids(
   const OrientationMap<VolumeDim>& orientation_map_from_me_to_neighbors =
-      previous_neighbors_in_direction.orientation();
+      previous_neighbors_in_direction.orientation().value_or(
+          OrientationMap<VolumeDim>::create_aligned());
```
This is used all over the code, can you keep it as a non-optional that is just `aligned`? That's what the missing value means anyway, so this change just makes the calling code more complicated.

kidder replied:
I did this on purpose to identify places that will need to depend upon InterfaceDataPolicy instead of the OrientationMap in order to support non-conforming blocks.  An OrientationMap has no meaning for the non-conforming cases as the logical coordinate axes on the interface are not simply related.

nilsvu replied:
Ok how about this then:

```cpp
const OrientationMap<Dim>& orientation() {
  ASSERT(interface_data_policy == CopyProject, "Orientation only has meaning if the interface data policy is CopyProject");
  return orientation_.value_or(OrientationMap<Dim>::create_aligned());
}
```

nilsdeppe replied:
I don't have a strong preference one way or the other. I agree that having to use `value_or` in a bunch of places is less readable. I don't see how changing to the `ASSERT`-based approach would make using the code more error prone than people misusing `value_or` to begin with in this case. I expect the suggestion will actually be safer than the current code because the class itself verifies consistency.

kidder replied:
okay, implemented the ASSERT based approach

nilsvu commented on src/Domain/Amr/UpdateAmrDecision.cpp:
```diff
@@ -40,7 +40,9 @@ bool update_amr_decision(
-          neighbors_in_dir.orientation();
+          neighbors_in_dir.orientation().value_or(
+              OrientationMap<VolumeDim>::create_aligned());
+      ;
```
extra semicolon

kidder replied:
oops

nilsvu commented on src/Domain/DomainHelpers.cpp:
```diff
@@ -460,11 +460,18 @@ void set_internal_boundaries(
+              std::move((orientation_helper.first)[0]),
+              Neighbors<VolumeDim, size_t>(
+                  {block2_index},
+                  domain::InterfaceDataPolicy::OrientCopyProject,
```
It seems that `OrientCopyProject` is just `CopyProject and not orientation.is_aligned()`. That's how we already handle these interfaces (checking if the orientation is aligned before orientating), so I'd remove `OrientCopyProject` and just keep doing the conditional orientation. The extra enum case just has the potential to be inconsistent with the value of `orientation.is_aligned()`.

kidder replied:
The OrientationMap will only be used for OrientCopyProject.  Instead of checking orientation.is_aligned(), code will need to execute a switch statement based on the InterfaceDataPolicy.

nilsvu replied:
Ok, can we do this (pseudocode):

```cpp
switch (neighbors.interface_data_policy) {
  case CopyProject:
    if (neigbors.orientation().is_aligned()) {
      orient();
    }
    project();
  // don't need separate OrientCopyProject case
  case Nonconforming:  // etc
    // don't need orientation
    interpolate();
```

nilsdeppe replied:
I don't see the benefit of the suggestion. I find it clearer with the enum describing what to do. I find the pseudo code:
```cpp
switch (neighbors.interface_data_policy) {
  case CopyProject:
    project();
  case OrientCopyProject:
    project(orient());
  case Nonconforming: 
    interpolate();
```
a lot clearer and self-documenting.

nilsvu commented on src/Domain/Structure/Neighbors.hpp:
```diff
@@ -90,8 +104,12 @@ class Neighbors {
-  OrientationMap<VolumeDim> orientation_;
+  domain::InterfaceDataPolicy interface_data_policy_{
+      domain::InterfaceDataPolicy::Uninitialized};
+  std::optional<OrientationMap<VolumeDim>> orientation_{std::nullopt};
```
Just make this non-optional and default-construct to aligned (or not default-construct at all, like currently). Has the same meaning and is easier to work with. I understand that you don't store the orientation map at all if you want to do interpolation between the faces, but I wonder if we want to do 1d interpolation at some point, and then we need the orientation anyway. I'd just keep the orientation here instead of making it an optional.

kidder replied:
I'm not sure what you mean by this.  Interpolation is done on the boundary (so a 2d interpolation for a 3d domain).  There is no relation between the logical axes tangent on each side of the interface that can be represented by an OrientationMap

nilsvu replied:
Makes sense, then the assert when calling `orientation()` would be the easiest to work with I think (see comment above). To avoid the `value_or` you could default-construct an aligned map here.

nilsvu commented on src/Domain/Structure/Neighbors.hpp:
```diff
@@ -34,29 +35,29 @@ class Neighbors {
   /// coordinate frame of the host Element and maps them to the logical
   /// coordinate frame of the neighbor Element.
-  Neighbors(std::unordered_set<ElementId<VolumeDim>> ids,
+  Neighbors(std::unordered_set<IdType> ids,
```
Suggestion: add a constructor with a single `IdType` value? Then you don't have to add curly braces in so many places.

nilsdeppe replied:
I can't think of a reason where that would subtly give unexpected results. As long as we document it, I think it'll work well :)

kidder replied:
done

nilsdeppe commented on src/Domain/Structure/Neighbors.cpp:
```diff
@@ -1,3 +1,4 @@
+
```
RIP

nilsvu replied:
ping on this

nilsdeppe commented on src/Domain/Structure/InterfaceDataPolicy.hpp:
```diff
@@ -0,0 +1,72 @@
+///   neighbor's boundary data and will interpolate it to the Mesh of the
+///   Element.  The Element will then need to send boundary correction data to
+///   the neighbor.
+/// - NonconformingNeighborInterpolates:  in this case the Element send its
```
`send` -> `sends`?

nilsdeppe commented on src/Domain/Structure/InterfaceDataPolicy.hpp:
```diff
@@ -0,0 +1,72 @@
+/// - For NonconformingSelfInterpolates and NonconformingNeighborInterpolates
+///   neighboring elements should have different values.  These cases should be
+///   used when one Element has many neighboring Elements (e.g. when a single
+///   spherical shell abuts a cubes sphere).  In this case it should be more
```
`cubes sphere` -> `cubed sphere`

nilsdeppe commented on src/Domain/Structure/InterfaceDataPolicy.hpp:
```diff
@@ -0,0 +1,72 @@
+///   NonconformingBothInterpolate, neighboring elements should agree on the
+///   values.
+/// - For NonconformingSelfInterpolates and NonconformingNeighborInterpolates
+///   neighboring elements should have different values.  These cases should be
```
maybe `different` -> `opposite`?

nilsdeppe commented on src/Domain/Structure/InterfaceDataPolicy.cpp:
```diff
@@ -0,0 +1,33 @@
+// Distributed under the MIT License.
```
Don't you need to add this to a CMakeLists.txt??

nilsdeppe reviewed (commented):
I like the overall design. I think Nils has some great suggestions. I've given my thoughts on this in general and some minor requests

kidder committed:
Add constructor for Neighbors taking a single id

This will allow fewer changes when converting BlockNeighbor into a
Neighbors.

Fixed ambiguous calls to construct Neighbors.

kidder committed:
Replace BlockNeighbor with strong typedef of Neighbors

Calls to BlockNeighbor::id() are replaced with calls to
BlockNeighbor::begin().

kidder committed:
Rename BlockNeighbor to BlockNeighbors

kidder commented:
- @nilsdeppe @nilsvu I implemented (most of) your suggestions.  
- In addition I kept the old constructors for Neighbors and set the InterfaceDataPolicy appropriately so as to touch less code in existing tests.  
- The first and fifth commits are essentially the same.  The others touched less existing code and are shorter (and I think less confusing) than fixup commits would have been.

nilsvu commented on src/Domain/Structure/Neighbors.hpp:
```diff
@@ -38,6 +38,15 @@ class Neighbors {
+  /// \param orientation This OrientationMap takes objects in the logical
+  /// coordinate frame of the host Element and maps them to the logical
+  /// coordinate frame of the neighbor Element.
+  Neighbors(const IdType& id, OrientationMap<VolumeDim> orientation);
```
As usual take `id` by value and move into new set?

nilsvu commented on src/Domain/CreateInitialElement.cpp:
```diff
@@ -43,7 +43,7 @@ Element<VolumeDim> create_initial_element(
         auto segment_ids_of_unrefined_neighbor = orientation(segment_ids);
         const auto& refinement_of_neighbor =
-            initial_refinement_levels[block_neighbor.id()];
+            initial_refinement_levels[*block_neighbor.begin()];
```
Assert that `block_neighbor.size() == 1` here and in other places you assume this?

nilsvu commented on src/Domain/Structure/Neighbors.cpp:
```diff
@@ -67,16 +68,37 @@ bool operator!=(const Neighbors<VolumeDim, IdType>& lhs,
-  p | orientation_;
+  if constexpr (std::is_same_v<IdType, size_t>) {
+    size_t version = 1;
+    p | version;
```
Add a comment that version 0 is deserializing the old `BlockNeighbor` class. That's what's happening here, right?

nilsvu commented on src/Domain/Structure/InterfaceDataPolicy.hpp:
```diff
@@ -0,0 +1,72 @@
+/// block logical coordinates of the neighboring Elements (Blocks) for the
+/// points on the interface between them.  In two cases there is a simple
+/// relationship between the coordinates:
+/// - CopyProject: in this case the block logical coordinates are identical.
```
This is a bit misleading, please clarify that the block logical coordinate _frame_ is the same, or something like that, but the two sides can still have different collocation points (different meshes).

nilsvu commented on src/Domain/Structure/Neighbors.cpp:
```diff
@@ -26,6 +26,14 @@ Neighbors<VolumeDim, IdType>::Neighbors(std::unordered_set<IdType> ids,
+template <size_t VolumeDim, typename IdType>
+Neighbors<VolumeDim, IdType>::Neighbors(const IdType& id,
+                                        OrientationMap<VolumeDim> orientation)
+    : ids_(std::unordered_set{id}), orientation_(std::move(orientation)) {
```
I think you can just delegate to the unordered_set constructor here, no need to reimplement it

nilsvu commented on src/Domain/Structure/Neighbors.hpp:
```diff
@@ -98,9 +118,19 @@ class Neighbors {
+                         const Neighbors<LocalVolumeDim, LocalIdType>& rhs);
+
+  std::unordered_set<IdType> ids_{};
+  domain::InterfaceDataPolicy interface_data_policy_{
```
I'm a bit hesitant about storing the interface data policy here because it couples details of our DG scheme into the domain structure. The domain structure doesn't need to know about things like projecting and interpolating. What do you think about storing this interface data policy in an option tag instead, which goes into the discretization group? We probably want to do the same policy on all cube/sphere boundaries anyway, so there's no need to store it separately at each interface. Then I think on the domain side of things we just need to store a `domain::BlockGeometry` enum with cases `Cube` and `SphericalShell` to have all the information we need.

kidder replied:
- It has nothing to do with DG specifically.  It just enables a switch based on how the coordinates on the two sides of the interface are related.
- The information needs to be in the Element just like the orientation as that is what is available at the decision points.
- No there is not a global value for the policy.  It is boundary dependent.  Especially as we are adding cylinders as well.
- There is no need for a BlockGeometry enum as the Basis and Quadrature already encode the information.

nilsvu replied:
Just to quickly clarify, I mean there's a global value for how cube/sphere interfaces are handled, and another global value for how cylinder interfaces are handled. We don't want to make this decision differently at each interface of the same type.

kidder commented:
@nilsvu I pushed fixups, please look at

nilsvu commented on src/Domain/Block.cpp:
```diff
@@ -23,7 +23,7 @@ Block<VolumeDim>::Block(
         Frame::BlockLogical, Frame::Inertial, VolumeDim>>&& stationary_map,
     const size_t id,
-    DirectionMap<VolumeDim, BlockNeighbor<VolumeDim>> neighbors,
+    DirectionMap<VolumeDim, BlockNeighbors<VolumeDim>> neighbors,
```
When you rebase you can assert here that the number of neighbors in each dimension is consistent with the topology of the block

kidder replied:
this is done in the next PR which also correctly sets the number of external boundaries based on the topology and input neighbors

nilsvu commented:
gcc9 has a complaint (maybe a missing include?)

kidder committed:
fix include order
